socorro/external/postgresql/connection_context.py

```diff
@@ -2,15 +2,32 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
+import os
 import socket
 import contextlib
 import psycopg2
 import psycopg2.extensions
+from urlparse import urlparse
 
-from configman.config_manager import RequiredConfig
-from configman import Namespace
+from configman import RequiredConfig, Namespace
+from configman.memoize import memoize
 
+#------------------------------------------------------------------------------
+# we're going to end up parsing the same url several times, there's no need to
+# actually do it more than once, so we wrap the method with a memoizer to speed
+# things along.
+@memoize()
```

Comment:
Solution 1 would be functionally equivalent to what I have now.  So it is a matter of style.  I prefer my method because the action of determining the default to use is encapsulated in a function that could be easily replaced.  I'm amused by your objection to the method not being _tight_ - if _tight_ means that it cannot use global state, then a _tight_ function could never use the environment.  An alternative method that would allow environment to be passed, as in dependency injection, in would be needless complication to the function call.  The only reason that your code proposal doesn't suffer the same criticism is the fact that it isn't encapsulated into a function.  

Solution 2 - I have several objections to that method:

I implemented such a thing on Monday as a separate PG ConnectionContext class with the idea that through config you could choose to use either class depending on how you wanted to set the PG credentials.  That method however, did nothing to address the problem that the environment variable ```database_url``` does not map in configman to our database credentials (```resource.postgresql.database_url```)  So additional measures similar to what I did in my code here would still have to be done to get the environment variable's value into the right configman option.  

Then, in order to allow for overriding certain fields within the URL, I'd still have to have the configman options for each of the component parts.  At that point, I have to establish a overriding heirarchy between the __url__ and the component parts.  If the user were to override on the command line, the same value that is the default for ```database_name```, do I override it in the __url__?  How do I detect that the default from the user specified value.  This was flaw in the original PR as it did not establish a predictable overriding scheme.  In ```test.sh``` I found a place where the __url__ overrode the component parts default, but on the next line the opposite was true, the component parts overrode the __url__.  It was too confusing, so I opted for a method that was perfectly predictable.  

My final argument against that is the legacy configuration and code.  I wanted to minimize the size of the job both in the code in the PR and later during deployments.

Solution 3 - is what I believe I have implemented...